connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaOffsetBackingStore.java

```diff
@@ -231,5 +223,45 @@ public synchronized Void get(long timeout, TimeUnit unit) throws InterruptedExce
         }
     }
 
+    private class KafkaOffsetReadFuture extends ConvertingFutureCallback<Void, Map<ByteBuffer, ByteBuffer>> implements OffsetReadFuture {
+        private final Collection<ByteBuffer> keys;
+        private final AtomicBoolean completed;
+
+        public KafkaOffsetReadFuture(Collection<ByteBuffer> keys) {
+            super(null);
+            this.keys = keys;
+            this.completed = new AtomicBoolean(false);
+        }
+
+        @Override
+        public Map<ByteBuffer, ByteBuffer> convert(Void result) {
+            Map<ByteBuffer, ByteBuffer> values = new HashMap<>();
+            for (ByteBuffer key : keys)
+                values.put(key, data.get(key));
+            return values;
+        }
 
+        @Override
+        public void prematurelyComplete() {
+            log.debug("Prematurely completing offset read future");
+            onCompletion(null, null);
+        }
+
+        @Override
+        public Map<ByteBuffer, ByteBuffer> get() throws InterruptedException, ExecutionException {
+            log.trace("Blocking on offset read request");
+            return super.get();
```

Comment:
I think this would return stale values for a key. if this future was prematurely closed, then we might still return a stale values for a key. we should be careful that this doesn't break any of the existing contracts of OffsetReader.

Reply:
Yes, this is noted in the description. However, if we want to stick strictly to the API for the `OffsetStorageReader`, once option we have is to simply return an empty map to all callers blocked on the future.